text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,396 @@
+
+### Nested Saves / API Transactions / Websocket Support
+
+Many applications wish to create or update and save multiple records together. Achieving this in `ember-data` today is unweildy and has many
+ difficult edge cases: one of which is correctly matching data received back from the API to the newly created records already on the client.
+
+A similar edge case occurs when a newly created record is saved for the first time and prior to receiving the request response the same record
+  is recieved via another means (background polling, websocket subscription, etc.). We have no means of matching the record returned by the alternative means to that of the request, leading to a second cache entry being created and an error once the initial request completes.
+
+One solution has been to generate and assign `id`s for records on the client, but this is not always desireable. These scenarios are a major 
+ motivation for `lid` in the `json-api` spec. Users wishing to solve these cases would be able to serialize the `lid` of the `Identifier`
+ for a newly created record and reflect that `lid` back in any payloads send from their API for the session to correctly match the payloads to the record.
+
+### Better Cache Serialization & Improved Infra for Offline Support
+
+In order to enable users to achieve full offline support, or to serialize the store for transport across the wire (for example as an advanced 
+ fastboot rehydration mode) the entire state of the store needs to be serializable. This RFC introduces the foundation for mechanisms through
+ which this can be later achieved.
+
+----
+
+## Detailed design
+
+```typescript
+const IS_IDENTIFIER = Symbol('is-identifier');
+
+// provided for additional debuggability
+const DEBUG_CLIENT_ORIGINATED = Symbol('record-originated-on-client');
+const DEBUG_IDENTIFIER_BUCKET = Symbol('identifier-bucket');
+
+// json-api payload version
+export interface ResourceIdentifierObject {
+  lid?: string;
+  id: string;
+  type: string;
+}
+
+export interface Identifier {
+  [IS_IDENTIFIER]: true;
+  [DEBUG_IDENTIFIER_BUCKET]: string;
+  lid: string;
+}
+
+export interface RecordIdentifier extends Identifier {
+  id: string | null;
+  type: string;
+  [DEBUG_CLIENT_ORIGINATED]: boolean;
+}
+```
+
+Note: the referential stability (object reference) of all identifiers created by the `store` is guaranteed.
+ E.g. any data that results in the lookup of an identifier producing the same `lid` token will return the
+ same `Identifier` instance. This is useful for being able to use identifiers for either `Map` or `WeakMap`
+ cache solutions.
+
+### Buckets
+
+In an ideal world, the `lid` of each `Identifier` would be a `v4` [uuid](https://en.wikipedia.org/wiki/Universally_unique_identifier),
+ making it unique in all contexts. However, due to requirements around design flexibility and performance we are only requiring that 
+ `Identifiers` be unique _within their bucket_ for the data they are intended to reference.
```

Comment:
What about this requirement + multiple stores?  It's something that's been thought about and maybe deserves a few words of clarity in the RFC.

Reply:
Great suggestion, will add 👍

Reply:
Added